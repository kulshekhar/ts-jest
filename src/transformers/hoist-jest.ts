// take care of including ONLY TYPES here, for the rest use `ts`
import { LogContexts, LogLevels } from 'bs-logger'
import {
  Block,
  ExpressionStatement,
  Node,
  SourceFile,
  Statement,
  TransformationContext,
  Transformer,
  Visitor,
} from 'typescript'

import { ConfigSet } from '../config/config-set'

/**
 * What methods of `jest` should we hoist
 */
const HOIST_METHODS = ['mock', 'unmock', 'enableAutomock', 'disableAutomock', 'deepUnmock']

/**
 * @internal
 */
export const name = 'hoisting-jest-mock'
// increment this each time the code is modified
/**
 * @internal
 */
export const version = 1

/**
 * The factory of hoisting transformer factory
 *
 * @param cs Current jest configuration-set
 * @internal
 */
export function factory(cs: ConfigSet) {
  const logger = cs.logger.child({ namespace: 'ts-hoisting' })
  /**
   * Our compiler (typescript, or a module with typescript-like interface)
   * To access Program or TypeChecker, do: cs.tsCompiler.program or cs.tsCompiler.program.getTypeChecker()
   */
  const ts = cs.compilerModule

  function shouldHoistExpression(expression: Node): boolean {
    return (
      ts.isCallExpression(expression) &&
      ts.isPropertyAccessExpression(expression.expression) &&
      HOIST_METHODS.includes(expression.expression.name.text) &&
      ((ts.isIdentifier(expression.expression.expression) && expression.expression.expression.text === 'jest') ||
        shouldHoistExpression(expression.expression.expression))
    )
  }

  /**
   * Checks whether given node is a statement that we need to hoist
   *
   * @param node The node to test
   */
  function shouldHoistNode(node: Node): node is ExpressionStatement {
    return ts.isExpressionStatement(node) && shouldHoistExpression(node.expression)
  }

  /**
   * Create a source file visitor which will visit all nodes in a source file
   *
   * @param ctx The typescript transformation context
   * @param sf The owning source file
   */
  function createVisitor(ctx: TransformationContext, _: SourceFile) {
    /**
     * Current block level
     */
    let level = 0
    /**
     * List of nodes which needs to be hoisted, indexed by their owning level
     */
    const hoisted: Statement[][] = []
    /**
     * Called when we enter a block to increase the level
     */
    const enter = () => {
      level++
      // reuse arrays
      if (hoisted[level]) {
        hoisted[level].splice(0, hoisted[level].length)
      }
    }
    /**
     * Called when we leave a block to decrease the level
     */
    const exit = () => level--
    /**
     * Adds a node to the list of nodes to be hoisted in the current level
     *
     * @param node The node to hoist
     */
    const hoist = (node: Statement) => {
      if (hoisted[level]) {
        hoisted[level].push(node)
      } else {
        hoisted[level] = [node]
      }
    }
    /**
     * Our main visitor, which will be called recursively for each node in the source file's AST
     *
     * @param node The node to be visited
     */
    const visitor: Visitor = node => {
      // enter this level
      enter()

      // visit each child
      let resultNode = ts.visitEachChild(node, visitor, ctx)

      // check if we have something to hoist in this level
      if (hoisted[level] && hoisted[level].length) {
        // re-order children so that hoisted ones appear first
        // this is actually the main job of this transformer
        const hoistedStmts = hoisted[level]
        const otherStmts = (resultNode as Block).statements.filter(s => !hoistedStmts.includes(s))
        const newNode = ts.getMutableClone(resultNode) as Block
        newNode.statements = ts.createNodeArray([...hoistedStmts, ...otherStmts])
        resultNode = newNode
      }

      // exit the level
      exit()

      if (shouldHoistNode(resultNode)) {
        // hoist into current level
        hoist(resultNode as Statement)
      }

      // finally returns the currently visited node
      return resultNode
    }
    return visitor
  }

  // returns the transformer factory
  return (ctx: TransformationContext): Transformer<SourceFile> =>
    logger.wrap(
      { [LogContexts.logLevel]: LogLevels.debug, call: null },
      'visitSourceFileNode(): hoisting',
      (sf: SourceFile) => ts.visitNode(sf, createVisitor(ctx, sf)),
    )
}
