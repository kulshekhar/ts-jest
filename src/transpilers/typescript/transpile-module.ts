import path from 'node:path'

import ts from 'typescript'

import { TsJestDiagnosticCodes } from '../../utils/messages'

const barebonesLibContent = `/// <reference no-default-lib="true"/>
interface Boolean {}
interface Function {}
interface CallableFunction {}
interface NewableFunction {}
interface IArguments {}
interface Number {}
interface Object {}
interface RegExp {}
interface String {}
interface Array<T> { length: number; [n: number]: T; }
interface SymbolConstructor {
    (desc?: string | number): symbol;
    for(name: string): symbol;
    readonly toStringTag: symbol;
}
declare var Symbol: SymbolConstructor;
interface Symbol {
    readonly [Symbol.toStringTag]: string;
}`
const barebonesLibName = 'lib.d.ts'
let barebonesLibSourceFile: ts.SourceFile | undefined

const carriageReturnLineFeed = '\r\n'
const lineFeed = '\n'
function getNewLineCharacter(options: ts.CompilerOptions): string {
  switch (options.newLine) {
    case ts.NewLineKind.CarriageReturnLineFeed:
      return carriageReturnLineFeed
    case ts.NewLineKind.LineFeed:
    default:
      return lineFeed
  }
}

type ExtendedTsTranspileModuleFn = (input: string, transpileOptions: ts.TranspileOptions) => ts.TranspileOutput

/**
 * Copy source code of {@link ts.transpileModule} from {@link https://github.com/microsoft/TypeScript/blob/main/src/services/transpile.ts}
 * with extra modifications:
 * - Remove generation of declaration files
 * - Allow using custom AST transformers with the internal created {@link Program}
 */
const transpileWorker: ExtendedTsTranspileModuleFn = (input, transpileOptions) => {
  barebonesLibSourceFile ??= ts.createSourceFile(barebonesLibName, barebonesLibContent, {
    languageVersion: ts.ScriptTarget.Latest,
  })

  const diagnostics: ts.Diagnostic[] = []

  const options: ts.CompilerOptions = transpileOptions.compilerOptions
    ? // @ts-expect-error internal TypeScript API
      ts.fixupCompilerOptions(transpileOptions.compilerOptions, diagnostics)
    : {}

  // mix in default options
  const defaultOptions = ts.getDefaultCompilerOptions()
  for (const key in defaultOptions) {
    if (Object.hasOwn(defaultOptions, key) && options[key] === undefined) {
      options[key] = defaultOptions[key]
    }
  }

  // @ts-expect-error internal TypeScript API
  for (const option of ts.transpileOptionValueCompilerOptions) {
    // Do not set redundant config options if `verbatimModuleSyntax` was supplied.
    if (options.verbatimModuleSyntax && new Set(['isolatedModules']).has(option.name)) {
      continue
    }

    options[option.name] = option.transpileOptionValue
  }

  // transpileModule does not write anything to disk so there is no need to verify that there are no conflicts between input and output paths.
  options.suppressOutputPathCheck = true

  // Filename can be non-ts file.
  options.allowNonTsExtensions = true
  options.declaration = false
  options.declarationMap = false

  const newLine = getNewLineCharacter(options)
  // if jsx is specified then treat file as .tsx
  const inputFileName =
    transpileOptions.fileName ?? (transpileOptions.compilerOptions?.jsx ? 'module.tsx' : 'module.ts')
  // Create a compilerHost object to allow the compiler to read and write files
  const compilerHost: ts.CompilerHost = {
    getSourceFile: (fileName) => {
      // @ts-expect-error internal TypeScript API
      if (fileName === ts.normalizePath(inputFileName)) {
        return sourceFile
      }

      // @ts-expect-error internal TypeScript API
      return fileName === ts.normalizePath(barebonesLibName) ? barebonesLibSourceFile : undefined
    },
    writeFile: (name, text) => {
      if (path.extname(name) === '.map') {
        sourceMapText = text
      } else {
        outputText = text
      }
    },
    getDefaultLibFileName: () => barebonesLibName,
    useCaseSensitiveFileNames: () => false,
    getCanonicalFileName: (fileName) => fileName,
    getCurrentDirectory: () => '',
    getNewLine: () => newLine,
    fileExists: (fileName) => {
      return fileName.endsWith('package.json') ? ts.sys.fileExists(fileName) : fileName === inputFileName
    },
    readFile: (fileName) => {
      return fileName.endsWith('package.json') ? ts.sys.readFile(fileName) : ''
    },
    directoryExists: () => true,
    getDirectories: () => [],
  }

  const sourceFile = ts.createSourceFile(inputFileName, input, {
    languageVersion: options.target ?? ts.ScriptTarget.ESNext,
    impliedNodeFormat: ts.getImpliedNodeFormatForFile(
      inputFileName,
      /*packageJsonInfoCache*/ undefined,
      compilerHost,
      options,
    ),
    // @ts-expect-error internal TypeScript API
    setExternalModuleIndicator: ts.getSetExternalModuleIndicator(options),
    jsDocParsingMode: transpileOptions.jsDocParsingMode ?? ts.JSDocParsingMode.ParseAll,
  })
  if (transpileOptions.moduleName) {
    sourceFile.moduleName = transpileOptions.moduleName
  }

  if (transpileOptions.renamedDependencies) {
    // @ts-expect-error internal TypeScript API
    sourceFile.renamedDependencies = new Map(Object.entries(transpileOptions.renamedDependencies))
  }

  // Output
  let outputText: string | undefined
  let sourceMapText: string | undefined
  const inputs = [inputFileName]
  const program = ts.createProgram(inputs, options, compilerHost)

  if (transpileOptions.reportDiagnostics) {
    diagnostics.push(...program.getSyntacticDiagnostics(sourceFile))
  }

  diagnostics.push(...program.getOptionsDiagnostics())

  // Emit
  const result = program.emit(
    /*targetSourceFile*/ undefined,
    /*writeFile*/ undefined,
    /*cancellationToken*/ undefined,
    /*emitOnlyDtsFiles*/ undefined,
    transpileOptions.transformers,
  )

  diagnostics.push(...result.diagnostics)

  if (outputText === undefined) {
    diagnostics.push({
      category: ts.DiagnosticCategory.Error,
      code: TsJestDiagnosticCodes.Generic,
      messageText: 'No output generated',
      file: sourceFile,
      start: 0,
      length: 0,
    })
  }

  return { outputText: outputText ?? '', diagnostics, sourceMapText }
}

export const tsTranspileModule = transpileWorker
