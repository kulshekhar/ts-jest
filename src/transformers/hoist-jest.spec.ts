import { testing } from 'bs-logger'
import * as tsc from 'typescript'

import * as hoist from './hoist-jest'

const CODE_WITH_HOISTING_NO_JEST_GLOBALS = `
import React from 'react'
import Unmocked from '../__test_modules__/Unmocked'
import Mocked from '../__test_modules__/Mocked'
import a from '../__test_modules__/a'
import b from '../__test_modules__/b'
import c from '../__test_modules__/c'
import d from '../__test_modules__/d'
import f from '../__test_modules__/f'
import jestBackticks from '../__test_modules__/jestBackticks'

// The virtual mock call below will be hoisted above this \`require\` call.
const virtualModule = require('virtual-module')

// These will all be hoisted above imports
jest.unmock('react')
jest.deepUnmock('../__test_modules__/Unmocked')
jest.unmock('../__test_modules__/c').unmock('../__test_modules__/d')

let e;
(function () {
  const _getJestObj = 42;
  e = require('../__test_modules__/e').default
  // hoisted to the top of the function scope
  jest.unmock('../__test_modules__/e')
})()

jest.mock('../__test_modules__/f', () => {
  if (!global.CALLS) {
    global.CALLS = 0
  }
  global.CALLS++

  return {
    _isMock: true,
    fn: () => {
      // The \`jest.mock\` transform will allow require, built-ins and globals.
      const path = require('path')
      const array = new Array(3)
      array[0] = path.sep
      return jest.fn(() => array)
    },
  };
})
jest.mock(\`../__test_modules__/jestBackticks\`)
jest.mock('virtual-module', () => 'kiwi', {virtual: true})
// This has types that should be ignored by the out-of-scope variables check.
jest.mock('has-flow-types', () => (props: {children: mixed}) => 3, {
  virtual: true,
})

// These will not be hoisted
jest.unmock('../__test_modules__/a').dontMock('../__test_modules__/b')
jest.unmock('../__test_modules__/' + 'a')
jest.dontMock('../__test_modules__/Mocked')
{
  const jest = {unmock: () => {}};
  // Would error (used before initialization) if hoisted to the top of the scope
  jest.unmock('../__test_modules__/a')
}

// This must not throw an error
const myObject = {mock: () => {}}
myObject.mock('apple', 27)

// Variable names prefixed with \`mock\` (ignore case) should not throw as out-of-scope
const MockMethods = () => {}
jest.mock('../__test_modules__/f', () => MockMethods)

console.log(Unmocked)
console.log(Mocked)
console.log(a)
console.log(b)
console.log(c)
console.log(d)
console.log(e)
console.log(virtualModule)
console.log(jestBackticks)
`
const CODE_WITH_HOISTING_HAS_JEST_GLOBALS = `
  import a from '../__test_modules__/a'
  import b from '../__test_modules__/b'

  import {jest} from '@jest/globals'
  import {jest as aliasedJest} from '@jest/globals'
  import * as JestGlobals from '@jest/globals'

  import c from '../__test_modules__/c'
  import d from '../__test_modules__/d'

  // These will be hoisted above imports

  jest.unmock('../__test_modules__/a')
  aliasedJest.unmock('../__test_modules__/b')
  JestGlobals.jest.unmock('../__test_modules__/c')

  // These will not be hoisted above imports

  {
    const jest = {unmock: () => {}};
    jest.unmock('../__test_modules__/d');
  }

  console.log(a)
  console.log(b)
  console.log(c)
  console.log(d)
`

const logger = testing.createLoggerMock()
const createFactory = () => hoist.factory({ logger, compilerModule: tsc } as any)
const transpile = (source: string) => tsc.transpileModule(source, { transformers: { before: [createFactory()] } })

describe('hoisting', () => {
  it('should have correct signature', () => {
    expect(hoist.name).toBe('hoisting-jest-mock')
    expect(typeof hoist.version).toBe('number')
    expect(hoist.version).toBeGreaterThan(0)
    expect(typeof hoist.factory).toBe('function')
  })

  it.each([CODE_WITH_HOISTING_NO_JEST_GLOBALS, CODE_WITH_HOISTING_HAS_JEST_GLOBALS])(
    'should hoist correctly jest methods',
    (data) => {
      const out = transpile(data)

      expect(out.outputText).toMatchSnapshot()
    },
  )
})
