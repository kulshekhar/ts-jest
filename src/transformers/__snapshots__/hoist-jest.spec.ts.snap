// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hoisting should hoist correctly jest methods 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
// These will all be hoisted above imports
jest.unmock('react');
jest.deepUnmock('../__test_modules__/Unmocked');
jest.unmock('../__test_modules__/c').unmock('../__test_modules__/d');
jest.mock('../__test_modules__/f', function () {
    if (!global.CALLS) {
        global.CALLS = 0;
    }
    global.CALLS++;
    return {
        _isMock: true,
        fn: function () {
            // The \`jest.mock\` transform will allow require, built-ins and globals.
            var path = require('path');
            var array = new Array(3);
            array[0] = path.sep;
            return jest.fn(function () { return array; });
        },
    };
});
jest.mock(\\"../__test_modules__/jestBackticks\\");
jest.mock('virtual-module', function () { return 'kiwi'; }, { virtual: true });
// This has types that should be ignored by the out-of-scope variables check.
jest.mock('has-flow-types', function () { return function (props) { return 3; }; }, {
    virtual: true,
});
jest.unmock('../__test_modules__/' + 'a');
jest.mock('../__test_modules__/f', function () { return MockMethods; });
var Unmocked_1 = require(\\"../__test_modules__/Unmocked\\");
var Mocked_1 = require(\\"../__test_modules__/Mocked\\");
var a_1 = require(\\"../__test_modules__/a\\");
var b_1 = require(\\"../__test_modules__/b\\");
var c_1 = require(\\"../__test_modules__/c\\");
var d_1 = require(\\"../__test_modules__/d\\");
var jestBackticks_1 = require(\\"../__test_modules__/jestBackticks\\");
// The virtual mock call below will be hoisted above this \`require\` call.
var virtualModule = require('virtual-module');
var e;
(function () {
    // hoisted to the top of the function scope
    jest.unmock('../__test_modules__/e');
    var _getJestObj = 42;
    e = require('../__test_modules__/e').default;
})();
// These will not be hoisted
jest.unmock('../__test_modules__/a').dontMock('../__test_modules__/b');
jest.dontMock('../__test_modules__/Mocked');
{
    // Would error (used before initialization) if hoisted to the top of the scope
    jest.unmock('../__test_modules__/a');
    var jest = { unmock: function () { } };
}
// This must not throw an error
var myObject = { mock: function () { } };
myObject.mock('apple', 27);
// Variable names prefixed with \`mock\` (ignore case) should not throw as out-of-scope
var MockMethods = function () { };
console.log(Unmocked_1.default);
console.log(Mocked_1.default);
console.log(a_1.default);
console.log(b_1.default);
console.log(c_1.default);
console.log(d_1.default);
console.log(e);
console.log(virtualModule);
console.log(jestBackticks_1.default);
"
`;

exports[`hoisting should hoist correctly jest methods 2`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var globals_1 = require(\\"@jest/globals\\");
var globals_2 = require(\\"@jest/globals\\");
var JestGlobals = require(\\"@jest/globals\\");
// These will be hoisted above imports
globals_1.jest.unmock('../__test_modules__/a');
globals_2.jest.unmock('../__test_modules__/b');
JestGlobals.jest.unmock('../__test_modules__/c');
var a_1 = require(\\"../__test_modules__/a\\");
var b_1 = require(\\"../__test_modules__/b\\");
var c_1 = require(\\"../__test_modules__/c\\");
var d_1 = require(\\"../__test_modules__/d\\");
// These will not be hoisted above imports
{
    jest_1.unmock('../__test_modules__/d');
    var jest_1 = { unmock: function () { } };
}
console.log(a_1.default);
console.log(b_1.default);
console.log(c_1.default);
console.log(d_1.default);
"
`;
